Fix culture-dependent date formatting tests to work with different system locales

Krishanthaudayakumara opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Apply the label "bug" or "enhancement" as applicable. -->

## Description / Motivation
<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

### Issue
Tests in the Sitecore ASP.NET Core SDK were failing when run on systems with non-US locales (e.g., Danish locale). The failures occurred because the tests were using hardcoded US date format expectations while the `DateTagHelper` implementation correctly uses `CultureInfo.CurrentCulture` when no specific culture is provided.

### Root Cause
The test assertions were inconsistent with the actual implementation behavior:
- **Implementation**: Uses `CultureInfo.CurrentCulture` when no culture is specified (line 65 in DateTagHelper.cs)
- **Tests**: Expected hardcoded US format strings (e.g., "05/04/2012") regardless of system locale

This mismatch caused test failures when running on systems with different date formatting conventions (e.g., Danish locale uses "05-04-2012" format).

### Solution
Updated test assertions to dynamically generate expected values using the current culture instead of hardcoded strings:

1. **Unit Tests** (`DateTagHelperFixture.cs`):
   - Modified custom format tests to explicitly specify culture and generate expected values accordingly
   - Added parameterized tests for multiple cultures (en-US, da-DK, uk-UA)

2. **Integration Tests**:
   - `AllFieldTagHelpersFixture.cs`: Use `TestConstants.DateTimeValue.ToString("MM/dd/yyyy", CultureInfo.CurrentCulture)`
   - `DateFieldTagHelperFixture.cs`: Generate expected format strings dynamically
   - `RichTextFieldTagHelperFixture.cs`: Use culture-aware date parsing and formatting
   - `ViewFieldsBindingFixture.cs`: Parse dates with invariant culture and format with current culture

### Behavior
- Tests now pass consistently regardless of the system locale
- Implementation behavior remains unchanged (still correctly uses current culture)
- Test coverage is improved with explicit multi-culture testing

This fix ensures the test suite is portable across different regional settings while maintaining the correct functionality of culture-aware date formatting.

Fixes https://github.com/Sitecore/ASP.NET-Core-SDK/issues/34

## Testing

- [x] The Unit & Intergration tests are passing.
- [x] I have added the necessary tests to cover my changes.

## Terms
<!-- Place an X in the [] to check. -->

<!-- The Code of Conduct helps create a safe space for everyone. We require that everyone agrees to it. -->
- [x] I agree to follow this project's [Code of Conduct](CODE_OF_CONDUCT.md).

Krishanthaudayakumara committed:
Ensure culture-invariant date formatting for explicit formats and set test cultures

- Updated DateTagHelper to use CultureInfo.InvariantCulture when DateFormat is specified and no explicit Culture is set, ensuring consistent date output regardless of system locale.
- Explicitly set culture to en-US in RichTextFieldTagHelperFixture and other relevant integration test fixtures to make test expectations stable and independent of the developer's environment.
- Updated test assertions to use the correct culture where appropriate.
- Resolves test failures related to date formatting when running tests under non-US locales (e.g., da-DK).

Krishanthaudayakumara committed:
Changes on copilot review - use en-US culture for test in DateFieldTagHelperFixture.cs,  extracting the culture selection logic into a separate variable for better readability in DateTagHelper.cs

Krishanthaudayakumara commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Integration.Tests/Fixtures/TagHelpers/DateFieldTagHelperFixture.cs:
```diff
@@ -38,6 +39,13 @@ public DateFieldTagHelperFixture()
                 app.UseRouting();
+                app.UseRequestLocalization(options =>
+                {
+                    var culture = new CultureInfo("da-DK");
```
Made the change to use "en-US"

Krishanthaudayakumara committed:
Move the formatCulture - formattedDate logic inside the ternary operator

Krishanthaudayakumara committed:
Refactor culture selection logic with formatCulture

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/DateTagHelper.cs:
```diff
@@ -67,7 +67,9 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
-        string formattedDate = !string.IsNullOrWhiteSpace(DateFormat) ? field.Value.ToString(DateFormat, culture) : field.Value.ToString(culture);
+        string formattedDate = !string.IsNullOrWhiteSpace(DateFormat)
+            ? field.Value.ToString(DateFormat, !string.IsNullOrWhiteSpace(Culture) ? culture : CultureInfo.InvariantCulture)
+            : field.Value.ToString(culture);
```
Changed it to this - 

```
CultureInfo culture = !string.IsNullOrWhiteSpace(Culture) ? CultureInfo.CreateSpecificCulture(Culture) : CultureInfo.CurrentCulture;

CultureInfo formatCulture = !string.IsNullOrWhiteSpace(Culture) ? culture : CultureInfo.InvariantCulture;

string formattedDate = !string.IsNullOrWhiteSpace(DateFormat)
            ? field.Value.ToString(DateFormat, formatCulture)
            : field.Value.ToString(culture);
```

robearlam reviewed (changes requested):
Added a couple of inline comments, also you've made code changes in the `DateTagHelper.cs` but I'm not seeing any changes to the Unit Tests for that class. If you've changed functionality then there should be a related unit test added to ensure it continues to function as expected.

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/DateTagHelper.cs:
```diff
@@ -67,7 +67,11 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
 
         CultureInfo culture = !string.IsNullOrWhiteSpace(Culture) ? CultureInfo.CreateSpecificCulture(Culture) : CultureInfo.CurrentCulture;
 
-        string formattedDate = !string.IsNullOrWhiteSpace(DateFormat) ? field.Value.ToString(DateFormat, culture) : field.Value.ToString(culture);
+        CultureInfo formatCulture = !string.IsNullOrWhiteSpace(Culture) ? culture : CultureInfo.InvariantCulture;
+
+        string formattedDate = !string.IsNullOrWhiteSpace(DateFormat)
+            ? field.Value.ToString(DateFormat, formatCulture)
+            : field.Value.ToString(culture);
```
Thanks for the feedback Rob,
In that implementation - 
- culture - Used for general date formatting when no specific DateFormat is provided
- formatCulture - Used specifically when a custom DateFormat is provided

so,
When `Culture` property is set, both culture and `formatCulture` use the specified culture
When `Culture` property is not set, culture uses `CurrentCulture`, but `formatCulture` uses `InvariantCulture`

but I have realize this seems not be the correct logic to move forward, I tried to adjust hardcoded date strings to match with culture and make date field tag helper tests culture-aware for testing here today - 
https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/20/commits 

I will keep looking towards to complete flow to figure out the correct solution for this case and update this thread.

Krishanthaudayakumara committed:
fix: Update date formatting tests to handle culture-specific formats - Convert DateTagHelper unit tests to use InlineData pattern with multiple cultures

Krishanthaudayakumara committed:
REmove manual culture setting

Krishanthaudayakumara committed:
Revert "Refactor culture selection logic with formatCulture"

This reverts commit debfcdef43edad7bea571981d891e831434b8991.

Krishanthaudayakumara committed:
Revert "Move the formatCulture - formattedDate logic inside the ternary operator"

This reverts commit 30986a284e59f05fe97faef2bc95f3dda72b14ff.

Krishanthaudayakumara committed:
Revert "Changes on copilot review - use en-US culture for test in DateFieldTagHelperFixture.cs,  extracting the culture selection logic into a separate variable for better readability in DateTagHelper.cs"

This reverts commit 979e5889d22bbfc6c017ae2621bc24e271e51aaf.

Krishanthaudayakumara committed:
Revert "Ensure culture-invariant date formatting for explicit formats and set test cultures"

This reverts commit ccc68c7609ec54a3da6e47053be78ff92e35ce80.

Krishanthaudayakumara committed:
Undo the previous changes - Date format culture invariant

Krishanthaudayakumara committed:
Refactor:  use [InlineAutoNSubstituteData] attributes and receive dependency-injected instances

Krishanthaudayakumara committed:
Update date formatting tests to handle culture-specific formats - Inline Data Approach

To main PR - Update date formatting tests to handle culture-specific formats - Inline Data Approach

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Integration.Tests/Fixtures/Binding/ViewFieldsBindingFixture.cs:
```diff
@@ -75,7 +76,8 @@ public async Task SitecoreLayoutModelBinders_BindDataCorrectly()
             .Should().BeEmpty();
 
-        sectionNode.ChildNodes.First(n => n.Name.Equals("textarea", StringComparison.OrdinalIgnoreCase)).InnerText.Should().Contain("12/12/2019");
+        DateTime expectedDate = DateTime.Parse("12.12.19", CultureInfo.InvariantCulture);
```
Is there a reason the test is defaulting to `CultureInfo.InvariantCulture` whereas the TagHelper code itself defaults to `CultureInfo.CurrentCulture`?

Krishanthaudayakumara replied:
Yes Rob, The original hardcoded expectation `"12/12/2019"` was failing on non-US locales because the DateTagHelper uses `CultureInfo.CurrentCulture` by default. So I changed it into that mentioned change, define the `expectedDate` with `InvariantCulture ` and next line when asserting, pass the `CurrentCulture` of that `expectedDate`, to ensures reliable parsing in any culture.

robearlam replied:
Makes sense, thanks for clarification!

Krishanthaudayakumara committed:
Merge branch 'main' into fix/date-format-culture-invariant
